### ISSUE TITLE
pljava crashes on Debian sid/i386

### ISSUE THREAD

**Issue by ChristophBerg** [tada/pljava#177]
Debian unstable/i386 with openjdk-8-jdk-headless.
After setting the correct libjvm.so location (see #176), installing the examples jar makes pljava crash:
```
2018-09-18 11:40:43.937 CEST [9528] cbe@postgres ANWEISUNG:  SELECT sqlj.install_jar('file:/usr/share/postgresql/10/pljava/pljava-examples-1.5.1-BETA2.jar', 'examples', true);
2018-09-18 11:40:55.939 CEST [9528] cbe@postgres WARNUNG:  Typeingabefunktion javatest.varlenaudttest_in sollte nicht VOLATILE sein
2018-09-18 11:40:55.939 CEST [9528] cbe@postgres ZUSAMMENHANG:  SQL-Anweisung »CREATE TYPE javatest.VarlenaUDTTest ( INPUT = javatest.VarlenaUDTTest_in, OUTPUT = javatest.VarlenaUDTTest_out, RECEIVE = javatest.VarlenaUDTTest_recv, SEND = javatest.VarlenaUDTTest_send, INTERNALLENGTH = VARIABLE, ALIGNMENT = INT4, STORAGE = PLAIN )«
2018-09-18 11:40:55.939 CEST [9528] cbe@postgres WARNUNG:  Typausgabefunktion javatest.varlenaudttest_out sollte nicht VOLATILE sein
2018-09-18 11:40:55.939 CEST [9528] cbe@postgres ZUSAMMENHANG:  SQL-Anweisung »CREATE TYPE javatest.VarlenaUDTTest ( INPUT = javatest.VarlenaUDTTest_in, OUTPUT = javatest.VarlenaUDTTest_out, RECEIVE = javatest.VarlenaUDTTest_recv, SEND = javatest.VarlenaUDTTest_send, INTERNALLENGTH = VARIABLE, ALIGNMENT = INT4, STORAGE = PLAIN )«
2018-09-18 11:40:55.939 CEST [9528] cbe@postgres WARNUNG:  Typempfangsfunktion javatest.varlenaudttest_recv sollte nicht VOLATILE sein
2018-09-18 11:40:55.939 CEST [9528] cbe@postgres ZUSAMMENHANG:  SQL-Anweisung »CREATE TYPE javatest.VarlenaUDTTest ( INPUT = javatest.VarlenaUDTTest_in, OUTPUT = javatest.VarlenaUDTTest_out, RECEIVE = javatest.VarlenaUDTTest_recv, SEND = javatest.VarlenaUDTTest_send, INTERNALLENGTH = VARIABLE, ALIGNMENT = INT4, STORAGE = PLAIN )«
2018-09-18 11:40:55.939 CEST [9528] cbe@postgres WARNUNG:  Typsendefunktion javatest.varlenaudttest_send sollte nicht VOLATILE sein
2018-09-18 11:40:55.939 CEST [9528] cbe@postgres ZUSAMMENHANG:  SQL-Anweisung »CREATE TYPE javatest.VarlenaUDTTest ( INPUT = javatest.VarlenaUDTTest_in, OUTPUT = javatest.VarlenaUDTTest_out, RECEIVE = javatest.VarlenaUDTTest_recv, SEND = javatest.VarlenaUDTTest_send, INTERNALLENGTH = VARIABLE, ALIGNMENT = INT4, STORAGE = PLAIN )«
2018-09-18 11:40:56.135 CEST [9528] cbe@postgres WARNUNG:  18 Sep 18 09:40:56 org.postgresql.pljava.example.LoggerTest SetOfRecordTest not ok
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0xeb45eac6, pid=9528, tid=0xf4c999c0
#
# JRE version: OpenJDK Runtime Environment (8.0_181-b13) (build 1.8.0_181-8u181-b13-1-b13)
# Java VM: OpenJDK Server VM (25.181-b13 mixed mode linux-x86 )
# Problematic frame:
# V  [libjvm.so+0x5eaac6]
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/cbe/tmp/pg_virtualenv.gmF7bD/data/10/regress/hs_err_pid9528.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#
2018-09-18 11:40:56.247 CEST [8428] LOG:  Serverprozess (PID 9528) wurde von Signal 6 beendet: Aborted
2018-09-18 11:40:56.247 CEST [8428] DETAIL:  Der fehlgeschlagene Prozess führte aus: SELECT sqlj.install_jar('file:/usr/share/postgresql/10/pljava/pljava-examples-1.5.1-BETA2.jar', 'examples', true);
2018-09-18 11:40:56.247 CEST [8428] LOG:  aktive Serverprozesse werden abgebrochen
```
[hs_err_pid9528.log](https://github.com/tada/pljava/files/2392283/hs_err_pid9528.log)

**Comment by jcflack** [comment]
Could you please try the following on this test host:

$ javap -c -private -classpath /path/to/pljava-1.5.1-BETA2.jar \
  org.postgresql.pljava.internal.VarlenaWrapper\$Input\$State | less

look for the disassembly of the first method:

    private org.postgresql.pljava.internal.VarlenaWrapper$Input$State(org.postgresql.pljava.internal.DualState$Key, org.postgresql.pljava.internal.VarlenaWrapper$Input, long, long, long, long, java.nio.ByteBuffer);

and confirm that offset 35 is:

    35: invokevirtual #7  // Method java/nio/ByteBuffer.asReadOnlyBuffer:()Ljava/nio/ByteBuffer;

(if it is not, please post what is at offset 35 in your build).

Then, please start a psql session, `CREATE EXTENSION pljava` if it is not already created, then reconnect with this psql command:

    \c "dbname=postgres options='-c pljava.vmoptions=-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=localhost:'"

(adjust `dbname=` and/or `user=` if needed for your setup.)

Request debug messages and wake pljava with a simple function call:

    SET client_min_messages TO debug1;
    SELECT sqlj.get_classpath('public');

Please post the resulting messages. They will begin with one like this:

    Listening for transport dt_socket at address: nnnnn

In another shell window, run:

    jdb -attach localhost:nnnnn
    stop at org.postgresql.pljava.internal.VarlenaWrapper$Input$State:236
    monitor locals
    monitor print buf

Back in the `psql` window, give the `SELECT sqlj.install_jar(...)` query to load the examples jar.

Please post what is displayed in the `jdb` window when the breakpoint is hit. Then give `jdb` the `cont` command.

If the crash does not follow immediately, the breakpoint will be hit again (there are a bunch of tests conducted, 48 I think), so just keep copying the displayed information and `cont`ing until the crash happens.

Thanks!

**Comment by jcflack** [comment]
The `javap` line above should be a single shell command; I used backslash continuation but forgot to mark it verbatim, and Markdown ate the backslash. (GitHub comment editing broken for me at the moment.)

**Comment by ChristophBerg** [comment]
```
$ /usr/lib/jvm/java-8-openjdk-i386/bin/javap -c -private -classpath debian/postgresql-10-pljava/usr/share/postgresql/10/pljava/pljava-1.5.1-BETA2.jar 'org.postgresql.pljava.internal.VarlenaWrapper$Input$State'
  private org.postgresql.pljava.internal.VarlenaWrapper$Input$State(org.postgresql.pljava.internal.DualState$Key, org.postgresql.pljava.internal.VarlenaWrapper$Input, long, long, long, long, java.nio.ByteBuffer);
    Code:
      35: invokevirtual #7                  // Method java/nio/ByteBuffer.asReadOnlyBuffer:()Ljava/nio/ByteBuffer;
```
```
\c "dbname=postgres options='-c pljava.vmoptions=-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=localhost:'"
set pljava.libjvm_location = '/usr/lib/jvm/java-8-openjdk-i386/jre/lib/i386/server/libjvm.so';
SELECT sqlj.get_classpath('public');
DEBUG:  00000: PL/Java loaded
DETAIL:  versions:
PL/Java native code (1.5.1-BETA2)
PL/Java common code (1.5.1-BETA2)
Built for (PostgreSQL 10.5 (Debian 10.5-2~30.git84cd21f.pgdg+1) on i686-pc-linux-gnu, compiled by gcc (Debian 8.2.0-4) 8.2.0, 32-bit)
Loaded in (PostgreSQL 10.5 (Debian 10.5-2~30.git84cd21f.pgdg+1) on i686-pc-linux-gnu, compiled by gcc (Debian 8.2.0-4) 8.2.0, 32-bit)
OpenJDK Runtime Environment (1.8.0_181-8u181-b13-1-b13)
OpenJDK Server VM (25.181-b13, mixed mode)
ORT:  initsequencer, Backend.c:597
```
There is no "Listening" line, but I got the port from `netstat -tlpn`. Unfortunately, the server dies on hitting the breakpoint:
```
$ /usr/lib/jvm/java-8-openjdk-i386/bin/jdb -attach localhost:46409
Set uncaught java.lang.Throwable
Set deferred uncaught java.lang.Throwable
Initializing jdb ...
> stop at org.postgresql.pljava.internal.VarlenaWrapper$Input$State:236
Set breakpoint org.postgresql.pljava.internal.VarlenaWrapper$Input$State:236
> monitor locals
> monitor print buf
> 
Breakpoint hit: "thread=main", org.postgresql.pljava.internal.VarlenaWrapper$Input$State.<init>(), line=236 bci=21

Internal exception:
com.sun.jdi.VMDisconnectedException
	at com.sun.tools.jdi.TargetVM.waitForReply(TargetVM.java:307)
	at com.sun.tools.jdi.VirtualMachineImpl.waitForTargetReply(VirtualMachineImpl.java:1030)
	at com.sun.tools.jdi.PacketStream.waitForReply(PacketStream.java:69)
	at com.sun.tools.jdi.JDWP$StackFrame$GetValues.waitForReply(JDWP.java:7106)
	at com.sun.tools.jdi.StackFrameImpl.getValues(StackFrameImpl.java:233)
	at com.sun.tools.example.debug.tty.Commands.commandLocals(Commands.java:1583)
	at com.sun.tools.example.debug.tty.TTY.executeCommand(TTY.java:447)
	at com.sun.tools.example.debug.tty.TTY.vmInterrupted(TTY.java:215)
	at com.sun.tools.example.debug.tty.EventHandler.run(EventHandler.java:81)
	at java.lang.Thread.run(Thread.java:748)
main[1] main[1] 
The application has been disconnected
```
```
# SELECT sqlj.install_jar('file:/usr/share/postgresql/10/pljava/pljava-examples-1.5.1-BETA2.jar', 'examples', true);
DEBUG:  00000: baue Index »pg_toast_16504_index« von Tabelle »pg_toast_16504«
ORT:  index_build, index.c:2031
DEBUG:  00000: baue Index »pg_toast_16518_index« von Tabelle »pg_toast_16518«
ORT:  index_build, index.c:2031
DEBUG:  00000: CREATE TABLE / PRIMARY KEY erstellt implizit einen Index »employees1_pkey« für Tabelle »employees1«
ORT:  DefineIndex, indexcmds.c:655
DEBUG:  00000: baue Index »employees1_pkey« von Tabelle »employees1«
ORT:  index_build, index.c:2031
DEBUG:  00000: CREATE TABLE / PRIMARY KEY erstellt implizit einen Index »employees2_pkey« für Tabelle »employees2«
ORT:  DefineIndex, indexcmds.c:655
DEBUG:  00000: baue Index »employees2_pkey« von Tabelle »employees2«
ORT:  index_build, index.c:2031
HINWEIS:  42809: Rückgabetyp javatest.udtscalariotest ist nur eine Hülle
ORT:  compute_return_type, functioncmds.c:105
HINWEIS:  42809: Argumenttyp javatest.udtscalariotest ist nur eine Hülle
ORT:  interpret_function_parameter_list, functioncmds.c:245
HINWEIS:  42809: Rückgabetyp javatest.udtscalariotest ist nur eine Hülle
ORT:  compute_return_type, functioncmds.c:105
HINWEIS:  42809: Argumenttyp javatest.udtscalariotest ist nur eine Hülle
ORT:  interpret_function_parameter_list, functioncmds.c:245
HINWEIS:  42809: Rückgabetyp javatest.varlenaudttest ist nur eine Hülle
ORT:  compute_return_type, functioncmds.c:105
HINWEIS:  42809: Argumenttyp javatest.varlenaudttest ist nur eine Hülle
ORT:  interpret_function_parameter_list, functioncmds.c:245
HINWEIS:  42809: Rückgabetyp javatest.varlenaudttest ist nur eine Hülle
ORT:  compute_return_type, functioncmds.c:105
HINWEIS:  42809: Argumenttyp javatest.varlenaudttest ist nur eine Hülle
ORT:  interpret_function_parameter_list, functioncmds.c:245
WARNUNG:  42P17: Typeingabefunktion javatest.varlenaudttest_in sollte nicht VOLATILE sein
ORT:  DefineType, typecmds.c:565
WARNUNG:  42P17: Typausgabefunktion javatest.varlenaudttest_out sollte nicht VOLATILE sein
ORT:  DefineType, typecmds.c:570
WARNUNG:  42P17: Typempfangsfunktion javatest.varlenaudttest_recv sollte nicht VOLATILE sein
ORT:  DefineType, typecmds.c:575
WARNUNG:  42P17: Typsendefunktion javatest.varlenaudttest_send sollte nicht VOLATILE sein
ORT:  DefineType, typecmds.c:580
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.management.Commands Type id = OID(600)
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
HINWEIS:  42809: Rückgabetyp javatest.complex ist nur eine Hülle
ORT:  compute_return_type, functioncmds.c:105
HINWEIS:  42809: Argumenttyp javatest.complex ist nur eine Hülle
ORT:  interpret_function_parameter_list, functioncmds.c:245
HINWEIS:  42809: Rückgabetyp javatest.complex ist nur eine Hülle
ORT:  compute_return_type, functioncmds.c:105
HINWEIS:  42809: Argumenttyp javatest.complex ist nur eine Hülle
ORT:  interpret_function_parameter_list, functioncmds.c:245
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.management.Commands Type id = OID(16624)
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
DEBUG:  00000: rehashing catalog cache id 63 for pg_transform; 33 tups, 16 buckets
ORT:  RehashCatCache, catcache.c:756
DEBUG:  00000: baue Index »pg_toast_16638_index« von Tabelle »pg_toast_16638«
ORT:  index_build, index.c:2031
DEBUG:  00000: baue Index »pg_toast_16644_index« von Tabelle »pg_toast_16644«
ORT:  index_build, index.c:2031
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.management.Commands Type id = OID(16664)
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.LoggerTest Looking good!
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
WARNUNG:  01000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.LoggerTest SetOfRecordTest not ok
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.LoggerTest Holdability OK
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.LoggerTest works for 32767
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.LoggerTest works for 32768
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.LoggerTest works for 65536
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.LoggerTest works for 1048576
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.annotation.Point pg_catalog.point from SQLInput
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.annotation.ComplexScalar javatest.complex from string
ZEILE 1: SELECT javatest.assertHasValues( CAST('(1,2)' AS javatest.co...
                                               ^
ANFRAGE:  SELECT javatest.assertHasValues( CAST('(1,2)' AS javatest.complex), 1, 2)
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.annotation.ComplexScalar javatest.complex to SQLOutput
ZEILE 1: SELECT javatest.assertHasValues( CAST('(1,2)' AS javatest.co...
                                               ^
ANFRAGE:  SELECT javatest.assertHasValues( CAST('(1,2)' AS javatest.complex), 1, 2)
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.annotation.ComplexScalar javatest.complex from SQLInput
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.annotation.ComplexTuple javatest.complextuple from SQLInput
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.annotation.ComplexTuple javatest.complextuple to SQLOutput
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
INFO:  00000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.annotation.ComplexTuple javatest.complextuple from SQLInput
ORT:  Java_org_postgresql_pljava_internal_Backend__1log, Backend.c:1621
DEBUG:  00000: rehashing catalog cache id 42 for pg_proc; 257 tups, 128 buckets
ORT:  RehashCatCache, catcache.c:756
DEBUG:  00000: rehashing catalog cache id 43 for pg_proc; 257 tups, 128 buckets
ORT:  RehashCatCache, catcache.c:756
DEBUG:  00000: rehashing catalog cache id 42 for pg_proc; 513 tups, 256 buckets
ORT:  RehashCatCache, catcache.c:756
DEBUG:  00000: rehashing catalog cache id 43 for pg_proc; 513 tups, 256 buckets
ORT:  RehashCatCache, catcache.c:756
Server beendete die Verbindung unerwartet
	Das heißt wahrscheinlich, dass der Server abnormal beendete
	bevor oder während die Anweisung bearbeitet wurde.
```
(I should really switch to English messages next time.)

**Comment by ChristophBerg** [comment]
FWIW, it also crashes with openjdk 10, `set pljava.libjvm_location = '/usr/lib/jvm/java-10-openjdk-i386/lib/server/libjvm.so';`.

**Comment by ChristophBerg** [comment]
I'll give the debugging another try tomorrow when I'm back in the company, I have the session still open.

**Comment by jcflack** [comment]
If it turns out that `print buf` triggers the crash, it's possible that `dump buf` instead (on the next try) might reveal something without triggering it.

**Comment by ChristophBerg** [comment]
I didn't manage to extract anything using jdb or gdb. jdb (both 8 and 10) just exit, maybe because the PG backend is crashing. It doesn't matter if "print buf", "dump buf", or "locals" are requested.
```
$ jdb -attach localhost:38827
Set uncaught java.lang.Throwable
Set deferred uncaught java.lang.Throwable
Initializing jdb ...
> stop at org.postgresql.pljava.internal.VarlenaWrapper$Input$State:236
Set breakpoint org.postgresql.pljava.internal.VarlenaWrapper$Input$State:236
> 2018-09-21 11:25:06.022 CEST [10701] cbe@postgres WARNUNG:  Typeingabefunktion javatest.varlenaudttest_in sollte nicht VOLATILE sein
2018-09-21 11:25:06.022 CEST [10701] cbe@postgres ZUSAMMENHANG:  SQL-Anweisung »CREATE TYPE javatest.VarlenaUDTTest ( INPUT = javatest.VarlenaUDTTest_in, OUTPUT = javatest.VarlenaUDTTest_out, RECEIVE = javatest.VarlenaUDTTest_recv, SEND = javatest.VarlenaUDTTest_send, INTERNALLENGTH = VARIABLE, ALIGNMENT = INT4, STORAGE = PLAIN )«
2018-09-21 11:25:06.022 CEST [10701] cbe@postgres WARNUNG:  Typausgabefunktion javatest.varlenaudttest_out sollte nicht VOLATILE sein
2018-09-21 11:25:06.022 CEST [10701] cbe@postgres ZUSAMMENHANG:  SQL-Anweisung »CREATE TYPE javatest.VarlenaUDTTest ( INPUT = javatest.VarlenaUDTTest_in, OUTPUT = javatest.VarlenaUDTTest_out, RECEIVE = javatest.VarlenaUDTTest_recv, SEND = javatest.VarlenaUDTTest_send, INTERNALLENGTH = VARIABLE, ALIGNMENT = INT4, STORAGE = PLAIN )«
2018-09-21 11:25:06.022 CEST [10701] cbe@postgres WARNUNG:  Typempfangsfunktion javatest.varlenaudttest_recv sollte nicht VOLATILE sein
2018-09-21 11:25:06.022 CEST [10701] cbe@postgres ZUSAMMENHANG:  SQL-Anweisung »CREATE TYPE javatest.VarlenaUDTTest ( INPUT = javatest.VarlenaUDTTest_in, OUTPUT = javatest.VarlenaUDTTest_out, RECEIVE = javatest.VarlenaUDTTest_recv, SEND = javatest.VarlenaUDTTest_send, INTERNALLENGTH = VARIABLE, ALIGNMENT = INT4, STORAGE = PLAIN )«
2018-09-21 11:25:06.022 CEST [10701] cbe@postgres WARNUNG:  Typsendefunktion javatest.varlenaudttest_send sollte nicht VOLATILE sein
2018-09-21 11:25:06.022 CEST [10701] cbe@postgres ZUSAMMENHANG:  SQL-Anweisung »CREATE TYPE javatest.VarlenaUDTTest ( INPUT = javatest.VarlenaUDTTest_in, OUTPUT = javatest.VarlenaUDTTest_out, RECEIVE = javatest.VarlenaUDTTest_recv, SEND = javatest.VarlenaUDTTest_send, INTERNALLENGTH = VARIABLE, ALIGNMENT = INT4, STORAGE = PLAIN )«
2018-09-21 11:25:06.238 CEST [10701] cbe@postgres WARNUNG:  21 Sep 18 09:25:06 org.postgresql.pljava.example.LoggerTest SetOfRecordTest not ok

Breakpoint hit: "thread=main", org.postgresql.pljava.internal.VarlenaWrapper$Input$State.<init>(), line=236 bci=21

main[1] locals
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0xeb394de3, pid=10701, tid=0x9aa7eb40
#
# JRE version: OpenJDK Runtime Environment (8.0_181-b13) (build 1.8.0_181-8u181-b13-1-b13)
# Java VM: OpenJDK Server VM (25.181-b13 mixed mode linux-x86 )
# Problematic frame:
# V  [libjvm.so+0x4eede3]
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/cbe/tmp/pg_virtualenv.8GJO2V/data/10/regress/hs_err_pid10701.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#
2018-09-21 11:25:14.467 CEST [32185] LOG:  Serverprozess (PID 10701) wurde von Signal 6 beendet: Aborted
2018-09-21 11:25:14.467 CEST [32185] DETAIL:  Der fehlgeschlagene Prozess führte aus: SELECT sqlj.install_jar('file:/usr/share/postgresql/10/pljava/pljava-examples-1.5.1-BETA2.jar', 'examples', true);
2018-09-21 11:25:14.467 CEST [32185] LOG:  aktive Serverprozesse werden abgebrochen
2018-09-21 11:25:14.468 CEST [10628] WARNUNG:  Verbindung wird abgebrochen wegen Absturz eines anderen Serverprozesses
2018-09-21 11:25:14.468 CEST [10628] DETAIL:  Der Postmaster hat diesen Serverprozess angewiesen, die aktuelle Transaktion zurückzurollen und die Sitzung zu beenden, weil ein anderer Serverprozess abnormal beendet wurde und möglicherweise das Shared Memory verfälscht hat.
2018-09-21 11:25:14.468 CEST [10628] TIPP:  In einem Moment sollten Sie wieder mit der Datenbank verbinden und Ihren Befehl wiederholen können.
Internal exception:
2018-09-21 11:25:14.469 CEST [10924] cbe@postgres FATAL:  das Datenbanksystem ist im Wiederherstellungsmodus
com.sun.jdi.VMDisconnectedException
2018-09-21 11:25:14.471 CEST [32185] LOG:  alle Serverprozesse beendet; initialisiere neu
	at jdk.jdi/com.sun.tools.jdi.TargetVM.waitForReply(TargetVM.java:310)

The application has been disconnected	at jdk.jdi/com.sun.tools.jdi.VirtualMachineImpl.waitForTargetReply(VirtualMachineImpl.java:1142)

	at jdk.jdi/com.sun.tools.jdi.PacketStream.waitForReply(PacketStream.java:87)
	at jdk.jdi/com.sun.tools.jdi.JDWP$StackFrame$GetValues.waitForReply(JDWP.java:7209)
	at jdk.jdi/com.sun.tools.jdi.StackFrameImpl.getValues(StackFrameImpl.java:244)
	at jdk.jdi/com.sun.tools.example.debug.tty.Commands.commandLocals(Commands.java:1582)
	at jdk.jdi/com.sun.tools.example.debug.tty.TTY.executeCommand(TTY.java:447)
	at jdk.jdi/com.sun.tools.example.debug.tty.TTY.<init>(TTY.java:781)
	at jdk.jdi/com.sun.tools.example.debug.tty.TTY.main(TTY.java:1083)
```
The gdb backtraces are useless and consist only of random garbage.

Is there anything else I could do to help here?

**Comment by jcflack** [comment]
Maybe. Instead of requesting all `locals` at once, maybe ask for them one by one:

```
print resourceOwner
print memContext
print snapshot
print varlenaPtr
print buf
```
My suspicion is that last one goes boom. The others might still contain some hints. This seems like it would have to be some width-of-datatypes issue because of the 32-bit architecture, but may aha! moment has not come yet.

**Comment by jcflack** [comment]
*my*

**Comment by jcflack** [comment]
Looking more closely at your logs, I do not see an answer yet, but I see what may be another reportable issue:

```
WARNUNG:  01000: 19 Sep 18 09:27:17 org.postgresql.pljava.example.LoggerTest SetOfRecordTest not ok
2018-09-21 11:25:06.238 CEST [10701] cbe@postgres WARNUNG:  21 Sep 18 09:25:06 org.postgresql.pljava.example.LoggerTest SetOfRecordTest not ok
```
Are you seeing that test `not ok` on any other platform?

It may be worth taking a moment to see what is happening there (which should probably be reported as a new issue). You could install the examples jar with `deploy => false` (so the `VarlenaWrapper` crash doesn't scotch it), and then explicitly declare the function involved in the test:

```sql
SELECT sqlj.set_classpath('public', 'examples');
CREATE OR REPLACE FUNCTION javatest.executeselecttorecords(
        selectSQL pg_catalog.varchar)
        RETURNS SETOF pg_catalog.RECORD
        LANGUAGE java VOLATILE
        AS 'org.postgresql.pljava.example.annotation.SetOfRecordTest.executeSelect(java.lang.String)';
```
and execute the equivalent test:

```sql
SELECT * FROM
javatest.executeselecttorecords('SELECT ''Foo'',  1,  1.5,  23.67,
''2005-06-01'',  ''20:56''::time, ''2006-02-04 23:55:10''') 
AS
r(t_varchar varchar, t_integer integer, t_float float, t_decimal decimal(8,2),
t_date date, t_time time, t_timestamp timestamp)
```

I assume what is happening is some inaccuracy in one (or more) of the returned values, but it will be interesting to see.

**Comment by jcflack** [comment]
But back to the crash issue. Beyond asking `jdb` for the locals individually:

```
print resourceOwner
print memContext
print snapshot
print varlenaPtr
print buf
```
It may well be useful to check some things with `gdb`. (Is `-Dso.debug=true` on the `mvn` command line?)

```
(gdb) break VarlenaWrapper.c:234
(gdb) cont

... in the psql window, do the install_jar ...
... there may be some innocent SIGSEGVs caught here that you can safely 'cont'
through; they're just the heap getting expanded ...

Breakpoint 1, pljava_VarlenaWrapper_Input
234	dbb = JNI_newDirectByteBuffer(VARDATA(vl), actual);
(gdb) step
JNI_newDirectByteBuffer
989	BEGIN_JAVA
(gdb) next
990	result = (*env)->NewDirectByteBuffer(env, address, capacity);
(gdb) p address
$5 = (void *) 0x28aab14
(gdb) p capacity
$6 = 324976
(gdb) next
991	END_JAVA
(gdb) p (*env)->GetDirectBufferAddress(env, result)
$7 = (void *) 0x28aab14
(gdb) p (*env)->GetDirectBufferCapacity(env, result)
$8 = 324976
```
The values of `address` and `capacity` above may be different from what you see, but the important thing is to retrieve them from the newly-constructed `ByteBuffer` and see that they match the values that were passed. In this example (from a working system) they do; your results may differ.

I am also looking at issue #152 and beginning to seriously wonder if there is a possibility the `nar-maven-plugin` is choosing `CFLAGS` that somehow differ from those that were used when building postgresql or openjdk. Do you have a record of the `CFLAGS` used to build those? (I know `pg_config --cflags` will report them for postgres; I assume you have a way to find out for openjdk.)

The `CFLAGS` being used by the `nar-maven-plugin` can be seen by adding `-X` on the `mvn` command line, piping the (torrential) output into `less`, and searching with `/^\[DEBUG\] gcc`
to find the line that matters.

**Comment by jcflack** [comment]
Do you think you might have some time to gather more of this information? I've got branches ready to pull for the rest of the issues you've identified, but this still perplexes me (as does the `SetOfRecordTest`).

**Comment by ChristophBerg** [comment]
Sorry for the delay, I got a bit exhausted by too many debugging threads in parallel.

Loading the jar without deploying it doesn't trigger any problem, nor does it hit the breakpoint:
```
# SELECT * FROM executeselecttorecords('SELECT ''Foo'',  1,  1.5,  23.67,
''2005-06-01'',  ''20:56''::time, ''2006-02-04 23:55:10''') 
AS                                                          
r(t_varchar varchar, t_integer integer, t_float float, t_decimal decimal(8,2),
t_date date, t_time time, t_timestamp timestamp);
 t_varchar │ t_integer │ t_float │ t_decimal │   t_date   │  t_time  │     t_timestamp     
───────────┼───────────┼─────────┼───────────┼────────────┼──────────┼─────────────────────
 Foo       │         1 │     1.5 │     23.67 │ 2005-06-01 │ 20:56:00 │ 2006-02-05 00:55:10
```

Now loading the examples.jar with deployment:
```
> stop at org.postgresql.pljava.internal.VarlenaWrapper$Input$State:236
Set breakpoint org.postgresql.pljava.internal.VarlenaWrapper$Input$State:236
> 
Breakpoint hit: "thread=main", org.postgresql.pljava.internal.VarlenaWrapper$Input$State.<init>(), line=236 bci=21

main[1] print resourceOwner
 resourceOwner = 1471688976
main[1] print memContext
 memContext = 1475894784
main[1] print snapshot
 snapshot = 0
main[1] print varlenaPtr
 varlenaPtr = 2527785008
main[1] print buf
com.sun.jdi.VMDisconnectedException
	at com.sun.tools.jdi.TargetVM.waitForReply(TargetVM.java:307)
	at com.sun.tools.jdi.VirtualMachineImpl.waitForTargetReply(VirtualMachineImpl.java:1030)
	at com.sun.tools.jdi.PacketStream.waitForReply(PacketStream.java:69)
	at com.sun.tools.jdi.JDWP$StackFrame$GetValues.waitForReply(JDWP.java:7106)
	at com.sun.tools.jdi.StackFrameImpl.getValues(StackFrameImpl.java:233)
	at com.sun.tools.jdi.StackFrameImpl.getValue(StackFrameImpl.java:201)
	at com.sun.tools.example.debug.expr.LValue$LValueLocal.getValue(LValue.java:435)
	at com.sun.tools.example.debug.expr.ExpressionParser.evaluate(ExpressionParser.java:84)

	at com.sun.tools.example.debug.tty.Commands.evaluate(Commands.java:114)The application has been disconnected

	at com.sun.tools.example.debug.tty.Commands.doPrint(Commands.java:1654)
	at com.sun.tools.example.debug.tty.Commands$3.action(Commands.java:1680)
	at com.sun.tools.example.debug.tty.Commands$AsyncExecution$1.run(Commands.java:66)
 buf = null
```
Thanks for the `-Dso.debug=true` hint, lintian (the .deb package linter) had already complained that there were no debugging symbols, but I had not paid attention to that yet. Will try gdb in a minute.

**Comment by ChristophBerg** [comment]
There are no other "not ok" in the PG log. (Besides the warnings that the type IO functions shouldn't be volatile.)

The values extracted by gdb match:
```
(gdb) break VarlenaWrapper.c:234
Breakpoint 1 at 0xebdac5b9: file /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/c/VarlenaWrapper.c, line 234.
(gdb) c
Continuing.

Thread 1 "postgres" received signal SIGSEGV, Segmentation fault.
0xdbfb5590 in ?? ()
(gdb) c
Continuing.

Thread 1 "postgres" received signal SIGSEGV, Segmentation fault.
0xdbfa75da in ?? ()
(gdb) c
Continuing.

Thread 11 "postgres" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x9b07eb40 (LWP 16709)]
0xdbfe8bbf in ?? ()
(gdb) c
Continuing.
[Switching to Thread 0xf4c999c0 (LWP 16691)]

Thread 1 "postgres" hit Breakpoint 1, pljava_VarlenaWrapper_Input (d=1493891196, parent=0x57e1362c, ro=0x57e75f34)
    at /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/c/VarlenaWrapper.c:234
234		dbb = JNI_newDirectByteBuffer(VARDATA(vl), actual);
(gdb) step
JNI_newDirectByteBuffer (address=0x96aaf034, capacity=324562)
    at /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/c/JNICalls.c:989
989		BEGIN_JAVA
(gdb) next
990		result = (*env)->NewDirectByteBuffer(env, address, capacity);
(gdb) p address
$1 = (void *) 0x96aaf034
(gdb) p capacity
$2 = 324562
(gdb) next
991		END_JAVA
(gdb) p (*env)->GetDirectBufferAddress(env, result)
$3 = (void *) 0x96aaf034
(gdb) p (*env)->GetDirectBufferCapacity(env, result)
$4 = 324562
```

CFLAGS:
* maven: `[DEBUG] gcc -Wno-sign-conversion -Wall -Wno-long-long -Wpointer-arith -Wconversion -c -g -fPIC -DPLJAVA_SO_VERSION=1.5.1-BETA2 '-DPLJAVA_LIBJVMDEFAULT="/usr/lib/jvm/java-8-openjdk-i386/jre/lib/i386/server/libjvm.so"' -DLinux -DGNU_GCC -I/home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava-so/target/nar/javah-include -I/usr/lib/jvm/java-8-openjdk-i386/include -I/usr/lib/jvm/java-8-openjdk-i386/include/linux -I/usr/include/postgresql -I/usr/include/postgresql/10/server -I/home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/include -I/home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava-so/target/nar/javah-include /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/c/type/Any.c /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/c/type/Date.c /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/c/HashMap.c /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/c/type/Long.c /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/c/SQLInputFromTuple.c /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/c/type/Time.c /home/cbe/projects/postgresql/pljava/postgresql-pljava/pljava-so/src/main/c/VarlenaWrapper.c`
* PG: `CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security` `CFLAGS_SL = -fPIC`
* openjdk-8: `CFLAGS := -W -Wall -Wno-unused -Wno-unused-parameter -Wno-parentheses -pipe -D_GNU_SOURCE -D_REENTRANT -D_LARGEFILE64_SOURCE -fno-omit-frame-pointer -D_LITTLE_ENDIAN -DLINUX -DNDEBUG -DARCH='"i586"' -Di586 -DRELEASE='"1.8.0_181"' -I/<<PKGBUILDDIR>>/build/jdk/include -I/<<PKGBUILDDIR>>/build/jdk/include/linux -I/<<PKGBUILDDIR>>/src/jdk/src/share/javavm/export -I/<<PKGBUILDDIR>>/src/jdk/src/solaris/javavm/export -I/<<PKGBUILDDIR>>/src/jdk/src/share/native/common -I/<<PKGBUILDDIR>>/src/jdk/src/solaris/native/common -Wdate-time -D_FORTIFY_SOURCE=2 -g -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -fno-delete-null-pointer-checks -fno-lifetime-dse -fPIC -I/<<PKGBUILDDIR>>/build/jdk/gensrc_headers` (no idea if I picked that correctly, better recheck the build log at https://buildd.debian.org/status/fetch.php?pkg=openjdk-8&arch=i386&ver=8u181-b13-1&stamp=1533392429&raw=0)

Did I answer all questions you had?

**Comment by jcflack** [comment]
```
# SELECT * FROM executeselecttorecords('SELECT ... ''2006-02-04 23:55:10''') 
AS  r(... timestamp timestamp);
 ... t_timestamp     
 ...─────────────────────
 ...2006-02-05 00:55:10
```

What time zone are you in (or, more to the point, what is `TimeZone` in your PG instance)?

**Comment by ChristophBerg** [comment]
CET aka Europe/Berlin.

**Comment by jcflack** [comment]
> Did I answer all questions you had?

I think that was the entire current crop.

So the C code is creating a perfectly serviceable `ByteBuffer` that knows its address and size when queried from C, and the address seems correct if it is `0x96aaf034` and the value of `varlenaPtr` visible in Java is `2527785008` or `0x96aaf030` pointing at the 4-byte header. Yet this apparently ordinary `ByteBuffer` becomes a `SEGV` bomb as soon as Java code makes any reference to it, and this is only on `i386`, but does happen both with `java-10-openjdk-i386` and `java-8-openjdk-i386`. Does that seem like a correct summary?

I know this is an increasing headache ... would you be willing to download an Oracle i386 Java 8 or 10 build into some temporary location, just to point `pljava.libjvm_location` at that and see what it does? Or one of the `AdoptOpenJDK` binary builds, for that matter?

-Chap

**Comment by ChristophBerg** [comment]
The summary is correct as far as I could grasp the issue (tbh I haven't tried *understanding* it yet).
I'll give Oracle-JDK a try soonish.

**Comment by jcflack** [comment]
>> What time zone are you in (or, more to the point, what is TimeZone in your PG instance)?
> CET aka Europe/Berlin.

Hey wait a minute. Please tell me the `SetOfRecordTest` `not ok` is happening for all of your architectures, not just i386 ....

**Comment by jcflack** [comment]
I looked through a bunch of the logs and only found `SetOfRecordTest ok` in the ones I looked at, whether 64 or 32 bit. So the difference in timestamps doesn't seem to be totally repeatable (though it did appear when you queried manually). The one-hour error very much suggests a summer-time calculation to me. I wonder if it somehow is affected by the current time when the query is performed? I believe it should not be (the actual timestamp being converted is used to determine the zone offset, AFAICT), but I do not have many alternate explanations.

Do you think the PG TimeZone is actually set for `CET` or for `Europe/Berlin`? According to `zdump`, `Europe/Berlin` would still be on `CEST` until October 28, is that right?

```
$ zdump -c 2018,2019 -v Europe/Berlin
Europe/Berlin  -9223372036854775808 = NULL
Europe/Berlin  -9223372036854689408 = NULL
Europe/Berlin  Sun Mar 25 00:59:59 2018 UT = Sun Mar 25 01:59:59 2018 CET isdst=0 gmtoff=3600
Europe/Berlin  Sun Mar 25 01:00:00 2018 UT = Sun Mar 25 03:00:00 2018 CEST isdst=1 gmtoff=7200
Europe/Berlin  Sun Oct 28 00:59:59 2018 UT = Sun Oct 28 02:59:59 2018 CEST isdst=1 gmtoff=7200
Europe/Berlin  Sun Oct 28 01:00:00 2018 UT = Sun Oct 28 02:00:00 2018 CET isdst=0 gmtoff=3600
Europe/Berlin  9223372036854689407 = NULL
Europe/Berlin  9223372036854775807 = NULL
```

**Comment by jcflack** [comment]
When time permits, could you please try an i386 build from the `bug/REL1_5_STABLE/issue177` branch?

**Comment by jcflack** [comment]
I think that all of the issues you have identified should be resolved in the `bug/REL1_5_STABLE/issue177` branch, but it would be especially helpful to have confirmation about this crash-on-i386 issue.

**Comment by ChristophBerg** [comment]
Sorry for the late answer. The TimeZone was really CET (inherited from my shell environment, I never bothered to change that to Europe/Berlin), and the 1h offset for the January timestamp is correct, if it was supposed to be converted to/from UTC.
The branch passes the regression tests set up in https://salsa.debian.org/postgresql/postgresql-pljava/blob/master/debian/tests/installjar, both on amd64 and i386.
Thanks!

**Comment by jcflack** [comment]
Thanks, I'll merge in that branch and mark this issue taken care of.

The timestamp-without-time-zone  conversion is _intended to_ preserve the supplied value unchanged, but because the original JDBC mapping is to `java.sql.Timestamp` which inherits from a class that has a time zone, it ends up, counterintuitively, depending on the PG session time zone and also, in this case, the Java installation default time zone. Because the end of that story is that it is better for code to use the JDBC 4.2 mapping to `java.time.LocalDateTime` instead, and the documentation and release notes already say so, I conclude that side issue is in fact not a bug. (The `SetOfRecordTest` is intended to verify that casts are applied when needed, not the correctness of the casts; I should probably update it to test with types other than `timestamp` just to avoid this particular surprise.)

**Comment by ChristophBerg** [comment]
As builds and tests now succeed, I took a snapshot of master at c676735a and updated the postgresql-pljava package on apt.postgresql.org with that. I'll put the packages into production shortly. Thanks!
Getting pljava back into Debian will need more work, depending on how many (build-)dependencies are still missing...

**Comment by jcflack** [comment]
Hold on ... I'm about to make a 1.5.1-BETA3. I'd prefer that you use that rather than an arbitrary snapshot (among other things, it makes the ALTER EXTENSION UPDATE stuff work better, should anyone want to use that.

**Comment by ChristophBerg** [comment]
Aye!

**Comment by ChristophBerg** [comment]
Put the packages live on apt.postgresql.org. Cheers!
